sway-lib-std/src/debug.sw

```diff
@@ -0,0 +1,2576 @@
+}
+
+pub struct DebugList {
+    f: Formatter,
+    has_entries: bool,
+}
+
+pub struct DebugTuple {
+    f: Formatter,
+    has_fields: bool,
+}
+
+pub struct Formatter {}
+
+impl Formatter {
+    pub fn print_newline(self) {
+        let lf = [10u8];
+        syscall_write(STDERR, __addr_of(lf), 1);
+    }
+
+    pub fn print_str(self, s: str) {
+        syscall_write(STDERR, s.as_ptr(), s.len());
+    }
+
+    pub fn print_u8(self, value: u8) {
+        let mut value = value;
+        let mut digits = [0u8; 64];
+        let mut i = 63;
+        while value > 0 {
+            let digit = value % 10;
+            digits[i] = digit + 48; // ascii zero = 48 
+            i -= 1;
+            value = value / 10;
+        }
+
+        syscall_write(STDERR, __addr_of(digits).add::<u8>(i), 64 - i);
+    }
+
+    pub fn print_u16(self, value: u16) {
+        let mut value = value;
+        let mut digits = [0u8; 64];
+        let mut i = 63;
+        while value > 0 {
+            let digit = asm(v: value % 10) {
+                v: u8
+            };
+            digits[i] = digit + 48; // ascii zero = 48 
+            i -= 1;
+            value = value / 10;
+        }
+
+        syscall_write(STDERR, __addr_of(digits).add::<u8>(i), 64 - i);
+    }
+
+    pub fn print_u32(self, value: u32) {
+        let mut value = value;
+        let mut digits = [0u8; 64];
+        let mut i = 63;
+        while value > 0 {
+            let digit = asm(v: value % 10) {
+                v: u8
+            };
+            digits[i] = digit + 48; // ascii zero = 48 
+            i -= 1;
+            value = value / 10;
+        }
+
+        syscall_write(STDERR, __addr_of(digits).add::<u8>(i), 64 - i);
+    }
+
+    pub fn print_u64(self, value: u64) {
+        let mut value = value;
+        let mut digits = [0u8; 64];
+        let mut i = 63;
+        while value > 0 {
+            let digit = asm(v: value % 10) {
+                v: u8
+            };
+            digits[i] = digit + 48; // ascii zero = 48 
+            i -= 1;
+            value = value / 10;
+        }
+
+        syscall_write(STDERR, __addr_of(digits).add::<u8>(i), 64 - i);
+    }
+
+    pub fn print_u256(self, value: u256) {
+        let mut value = value;
+        // u256::MAX = 115792089237316195423570985008687907853269984665640564039457584007913129639935
+        let mut digits = [0u8; 80];
+        let mut i = 79;
+        while value > 0 {
+            let rem = value % 10;
+            let (_, _, _, digit) = asm(rem: rem) {
+                rem: (u64, u64, u64, u64)
+            };
+            let digit = asm(v: digit % 10) {
+                v: u8
+            };
+            digits[i] = digit + 48; // ascii zero = 48 
+            i -= 1;
+            value = value / 10;
+        }
+
+        syscall_write(STDERR, __addr_of(digits).add::<u8>(i), 80 - i);
+    }
+
+    pub fn debug_struct(self, name: str) -> DebugStruct {
+        self.print_str(name);
+        self.print_str(" { ");
```

Comment:
I probably dont understand this; but don't we need closing braces for these complex types?

Reply:
The closing braces are done in the `finish` methods of the corresponding `DebugStruct/List/Tuple` structs. This is a builder pattern where `finish` does the last step including closing braces.